File: Weel_1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int count_digits(long long int number);
int first_n_digits(long long int number, int n);
bool is_valid(long long int number, int length);

int main(void)
{
    // user's input
    long long int number = get_long_long("Number: ");

    // input's length
    int length = count_digits(number);

    // first digit of the number
    int first_digit = first_n_digits(number, 1);

    // first two digits of the number
    int first_two_digits = first_n_digits(number, 2);

    // check if the length is appropriate
    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // initialize variable - the type of the card or "Invalid" (if it's neither AmericanExpress / MasterCard/ Visa)
    string card;

    // AmericanExpress
    if (length == 15)
    {
        if (first_two_digits == 34 || first_two_digits == 37)
        {
            card = "AMEX\n";
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }
    else
    {
        // VISA
        if (length == 13)
        {
            if (first_digit == 4)
            {
                card = "VISA\n";
            }
            else
            {
                printf("INVALID\n");
                return 0;
            }
        }
        else
        {
            // VISA or MasterCard
            if (length == 16)
            {
                // VISA
                if (first_digit == 4)
                {
                    card = "VISA\n";
                }
                else
                {
                    // MasterCard
                    if (first_two_digits >= 51 && first_two_digits <= 55)
                    {
                        card = "MASTERCARD\n";
                    }
                    // Invalid input
                    else
                    {
                        printf("INVALID\n");
                        return 0;
                    }
                }
            }

        }
    }

    // Luhn's algo
    if (is_valid(number, length))
    {
        printf("%s", card);
    }
    else
    {
        printf("INVALID\n");
    }
}

// create function that returns the length of the long long integer
int count_digits(long long int number)
{
    int digits = 0;
    do
    {
        ++digits;
        number /= 10;
    } while (number);
    return digits;
}

// create function that returns the first N digits of the long long integer
int first_n_digits(long long int number, int n)
{
    while (number >= pow(10, n))
    {
        number /= 10;
    }
    return number;
}

//  Luhn's algorithm to check if the a card number is syntactically valid
bool is_valid(long long int number, int length)
{
    // let’s first underline every other digit, starting with the number’s second-to-last digit, and then multiply
    // each of them by 2, and then sum these product's digits (not the products themselves)
    int multiplied_nums_sum = 0;
    long long int number_copy = number;
    number_copy /= 10;   // to start with the number's second digit
    int digit;
    int twiced_digit;
    while (number_copy > 0)
    {
        digit = number_copy % 10;
        twiced_digit = digit * 2;
        while (twiced_digit > 0)
        {
            int products_digit = twiced_digit % 10;
            twiced_digit /= 10;
            multiplied_nums_sum += products_digit;
        }
        number_copy = number_copy / 100;
    }

    // let's add those numbers that weren't multiplied by 2
    int not_multiplied_nums_sum = 0;
    number_copy = number;
    while (number_copy > 0)
    {
        digit = number_copy % 10;
        number_copy /= 100;
        not_multiplied_nums_sum += digit;
    }

    // let's add those two sums
    int total = multiplied_nums_sum + not_multiplied_nums_sum;

    // If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    int last_digit = total % 10;
    if (last_digit == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: Week_6/credit.py
from cs50 import get_int

#  Luhn's algorithm to check if the a card number is syntactically valid
def luhns_is_valid(number):
    # let’s first underline every other digit, starting with the number’s second-to-last digit
    # and multiply each of them by 2
    number_reversed = str(number)[::-1]
    second_to_last_digits = [int(i) * 2 for i in number_reversed[1::2]]
    # then sum these product's digits (not the products themselves)
    second_to_last_digits_sum = 0
    for num in second_to_last_digits:
        while num > 0:
            second_to_last_digits_sum += num % 10
            num = num // 10
    print(second_to_last_digits_sum)
    # let's add those numbers that weren't multiplied by 2
    other_digits = [int(i) for i in number_reversed[::2]]
    other_digits_sum = sum(other_digits)
    print(other_digits_sum)
    # let's add those two sums
    total = second_to_last_digits_sum + other_digits_sum
    # If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    last_digit = total % 10
    if last_digit == 0:
        return True
    else:
        return False


# get input card number from user
number = get_int("Number: ")

# how many digits are in card number
length = len(str(number))

# flag that describes if card number is valid
is_invalid = False

# check the length of card number
if length not in [13, 15, 16]:
    is_invalid = True

# get the 1st two digits of card number
first_two_digits = int(str(number)[:2])   # 123456789 --> 12
# get the 1st digit of card number
first_digit = int(str(number)[0])          # 123456789 --> 1


if length == 15:
    if first_two_digits == 34 or first_two_digits == 37:
        card = "AMEX\n"
    else:
        is_invalid = True

if length == 13:
    if first_digit == 4:
        card = "VISA\n"
    else:
        is_invalid = True

if length == 16:
    if first_digit == 4:
        card = "VISA\n"
    elif 51 <= first_two_digits <= 55:
        card = "MASTERCARD\n"
    else:
        is_invalid = True

# last check by Luhn's algo
if luhns_is_valid(number) == False:
    is_invalid = True

if is_invalid == True:
    card = "INVALID"

print(card)



